sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java

```diff
@@ -0,0 +1,250 @@
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.beam.sdk.io.hcatalog.HCatalogIO.BoundedHCatalogSource;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.SourceTestUtils;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.util.UserCodeException;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hive.hcatalog.data.DefaultHCatRecord;
+import org.apache.hive.hcatalog.data.HCatRecord;
+import org.apache.hive.hcatalog.data.transfer.ReaderContext;
+import org.hamcrest.Matchers;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TestRule;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+
+/**
+ *Test for HCatalogIO.
+ */
+public class HCatalogIOTest implements Serializable {
+
+  @Rule
+  public final transient TestPipeline pipeline = TestPipeline.create();
+
+  @Rule
+  public transient ExpectedException thrown = ExpectedException.none();
+
+  @Rule
+  public final transient TestRule testDataSetupRule = new TestWatcher() {
+    public Statement apply(final Statement base, final Description description) {
+      return new Statement() {
+        @Override
+        public void evaluate() throws Throwable {
+          if (description.getAnnotation(TestDataSetupReqd.class) != null) {
+            prepareTestData();
+          } else if (description.getAnnotation(TestDataInitializationReqd.class) != null) {
+            reCreateTestTable();
+          }
+          base.evaluate();
+        }
+      };
+    }
+  };
+
+  /**
+   * Use this annotation to setup complete test data(table populated with records).
+   */
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target({ElementType.METHOD})
+  @interface TestDataSetupReqd {
+  }
+
+  /**
+   * Use this annotation to setup test tables alone(empty tables, no records are populated).
+   */
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target({ElementType.METHOD})
+  @interface TestDataInitializationReqd {
+  }
+
+
+  /**
+   * Perform end-to-end test of Write-then-Read operation.
+   */
+  @Test
+  @TestDataInitializationReqd
+  public void testWriteWriteThenReadSuccess() throws Exception {
+    pipeline.apply(Create.of(getDefaultHCatRecords(TEST_RECORDS_COUNT)))
+    .apply(HCatalogIO.write()
+        .withConfigProperties(getConfigProperties())
+        .withTable(TEST_TABLE_NAME));
+
+    pipeline.run();
+
+    PCollection<String> output = pipeline.apply(HCatalogIO.read()
```

Comment:
Interesting it worked with same pipeline! I have modified to use two separate pipelines in this case.